File: golang-code/ex15/erro.go
package main

type SaldoInsuficienteError struct {
	Mensagem string
}

func (e *SaldoInsuficienteError) Error() string {
	return e.Mensagem
}

type ContaBancaria struct {
	saldo float64
}

func (c *ContaBancaria) Depositar(valor float64) {
	c.saldo += valor
}

func (c *ContaBancaria) Sacar(valor float64) error {
	if valor > c.saldo {
		return &SaldoInsuficienteError{"Saldo insuficiente para a operação"}
	}

	c.saldo -= valor
	return nil
}

func (c *ContaBancaria) ObterSaldo() float64 {
	return c.saldo
}

File: python-code/ex15/erro.py
class SaldoInsuficienteException(Exception):
  def __init__(self, mensagem="Saldo insuficiente para a operação"):
    self.mensagem = mensagem
    super().__init__(self.mensagem)

class ContaBancaria:
  def __init__(self, saldo):
    self.__saldo = saldo

  def depositar(self, valor):
    self.__saldo += valor

  def sacar(self, valor):
    if valor > self.__saldo:
      raise SaldoInsuficienteException()
    self.__saldo -= valor

  def obter_saldo(self):
    return self.__saldo

def main():
  conta = ContaBancaria(200)
  
  try:
    conta.sacar(300)
  except SaldoInsuficienteException as e:
    print(e)

if __name__ == "__main__":
  main()